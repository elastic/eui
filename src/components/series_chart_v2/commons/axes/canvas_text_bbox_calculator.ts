import { none, Option, some } from 'fp-ts/lib/Option';
import { BBox, BBoxCalculator } from './bbox_calculator';

export class CanvasTextBBoxCalculator implements BBoxCalculator {
  private attachedRoot: HTMLElement;
  private offscreenCanvas: HTMLCanvasElement;
  private context: CanvasRenderingContext2D | null;
  // TODO specify styles for text
  // TODO specify how to hide the svg from the current dom view
  // like moving it a -9999999px
  constructor(rootElement?: HTMLElement) {
    this.offscreenCanvas = document.createElement('canvas');
    this.context = this.offscreenCanvas.getContext('2d');
    this.attachedRoot = rootElement || document.documentElement;
    this.attachedRoot.appendChild(this.offscreenCanvas);
  }
  public compute(text: string, fontSize = 16, fontFamily = 'Arial'): Option<BBox> {
    if (!this.context) {
      return none;
    }
    this.context.font = `${fontSize}px ${fontFamily}`;
    const measure = this.context.measureText(text);
    return some({
      width: measure.width,
      height: fontSize,
    });
  }
  public destroy(): void {
    this.attachedRoot.removeChild(this.offscreenCanvas);
  }
}
