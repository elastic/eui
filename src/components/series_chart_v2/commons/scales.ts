import { scaleBand, ScaleContinuousNumeric, scaleLinear, scaleLog, scaleSqrt } from 'd3-scale';

export type ScaleFunction = (point: any) => number;

export const enum ScaleType {
  Linear = 'linear',
  Ordinal = 'ordinal',
  Log = 'log',
  Sqrt = 'sqrt',
}

export const SCALES = {
  [ScaleType.Linear]: scaleLinear,
  [ScaleType.Log]: scaleLog,
  [ScaleType.Sqrt]: scaleSqrt,
};

export interface ScaleConfig {
  accessor: (value: any) => any;
  domain: string[] | number[] | [number, number] | [undefined, undefined];
  type: ScaleType;
  clamp?: boolean;
}

export interface ScaleOrdinal {
  (value: string): number;
  domain(domain?: string[]): string[] | this;
  range(range?: [number, number]): [number, number] | this;
  ticks(): string[];
  bandwidth(): number;
}

export function createScaleOrdinal(): ScaleOrdinal {
  const d3ScaleBand = scaleBand();
  const scale = (value: string) => {
    return Number(d3ScaleBand(value)) + (d3ScaleBand.bandwidth() / 2);
  };
  const ordinalScale = scale as ScaleOrdinal;
  ordinalScale.domain = (domain?: string[]): string[] | ScaleOrdinal => {
    if (domain) {
      d3ScaleBand.domain(domain);
      return ordinalScale;
    }
    return d3ScaleBand.domain();
  };
  ordinalScale.range = (range: [number, number]) => {
    if (range) {
      d3ScaleBand.range(range);
      return ordinalScale;
    }
    return d3ScaleBand.range();
  };
  ordinalScale.ticks = (): string[] => {
    return d3ScaleBand.domain();
  };
  ordinalScale.bandwidth = (): number => {
    return d3ScaleBand.bandwidth();
  };
  return ordinalScale;
}

type Scales = ScaleContinuousNumeric<number, number> | ScaleOrdinal;
export type ScaleContinuousTypes = ScaleType.Linear | ScaleType.Sqrt | ScaleType.Log;
export type ScaleOrdinalTypes =  ScaleType.Ordinal;
export type ScaleTypes = ScaleContinuousTypes | ScaleOrdinalTypes;

export function getScaleFromType(type: ScaleContinuousTypes): ScaleContinuousNumeric<number, number>;
export function getScaleFromType(type: ScaleOrdinalTypes): ScaleOrdinal;
export function getScaleFromType(type: ScaleTypes): Scales {
  // TODO add clamping
  if (type === ScaleType.Ordinal) {
    return createScaleOrdinal();
  }
  return SCALES[type]();
}

/**
 *
 * @param type {ScaleContinuousTypes} The type of continuous scale
 * @param domain The domain of the scale
 * @param accessor The accessor of the datum
 * @param minRange The min range of the scale
 * @param maxRange The max range of the scale
 */
export function getContinuousScaleFn(
  type: ScaleContinuousTypes,
  domain: number[],
  accessor: (point: any) => any,
  minRange: number,
  maxRange: number,
  clamp: boolean = false,
): ScaleFunction {
  const scale = createContinuousScale(type, domain, minRange, maxRange, clamp);
  return (point: any) => {
    return scale(accessor(point));
  };
}

export function createContinuousScale(
  type: ScaleContinuousTypes,
  domain: number[],
  minRange: number,
  maxRange: number,
  clamp: boolean = false,
): ScaleContinuousNumeric<number, number> {
  const scale = getScaleFromType(type);
  scale.domain(domain);
  scale.range([minRange, maxRange]);
  scale.clamp(clamp);
  return scale;
}

/**
 * Return a ScaleOrdinal
 * @param domain The domain of the ordinal scale
 * @param minRange The max range of the scale
 * @param maxRange The min range of the scale
 */
export function createOrdinalScale(
  domain: string[],
  minRange: number,
  maxRange: number,
): ScaleOrdinal {
  const scale = getScaleFromType(ScaleType.Ordinal);
  scale.domain(domain);
  scale.range([minRange, maxRange]);
  return scale;
}

/**
 * Instanciate an ordinal scale function
 * @param scale The ordinal Scale
 * @param accessor The datum accessor function
 * @returns {ScaleFunction} The ScaleFunction
 */
export function getOrdinalScaleFn(
  scale: ScaleOrdinal,
  accessor: (point: any) => any,
): ScaleFunction {
  return (point: any) => {
    return scale(accessor(point));
  };
}
