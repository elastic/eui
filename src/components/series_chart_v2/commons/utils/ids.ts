import { iso, Newtype } from 'newtype-ts';

export interface GroupId extends Newtype<{ readonly GroupId: unique symbol }, string> {}
export interface AxisId extends Newtype<{ readonly AxisId: unique symbol }, string> {}
export interface SpecId extends Newtype<{ readonly SpecId: unique symbol }, string> {}
const groupIdIso = iso<GroupId>();
const axisIdIso = iso<AxisId>();
const specIdIso = iso<SpecId>();

export function getGroupId(id: string): GroupId {
  return groupIdIso.wrap(id);
}
export function getAxisId(id: string): AxisId {
  return axisIdIso.wrap(id);
}
export function getSpecId(id: string): SpecId {
  return specIdIso.wrap(id);
}
