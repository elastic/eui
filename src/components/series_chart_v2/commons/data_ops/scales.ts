import { scaleBand, ScaleContinuousNumeric, scaleLinear, scaleLog, scaleSqrt } from 'd3-scale';
import { Datum } from '../series/specs';
import { AccessorFn } from './accessor';

export type ScaleFn = (datum: Datum) => number;

export const enum ScaleType {
  Linear = 'linear',
  Ordinal = 'ordinal',
  Log = 'log',
  Sqrt = 'sqrt',
}

const SCALES = {
  [ScaleType.Linear]: scaleLinear,
  [ScaleType.Log]: scaleLog,
  [ScaleType.Sqrt]: scaleSqrt,
};

export interface ScaleConfig {
  accessor: (value: any) => any;
  domain: string[] | number[] | [number, number] | [undefined, undefined];
  type: ScaleType;
  clamp?: boolean;
}

export interface ScaleOrdinal {
  (value: string): number;
  domain: (domain?: string[]) => string[] | this;
  range: (range?: [number, number]) => [number, number] | this;
  ticks(): string[];
  bandwidth(): number;
}

export type ScaleContinuousTypes = ScaleType.Linear | ScaleType.Sqrt | ScaleType.Log;
export type ScaleOrdinalTypes = ScaleType.Ordinal;
export type ScaleTypes = ScaleContinuousTypes | ScaleOrdinalTypes;

export type ScaleFunction = (value: Datum) => any;

function buildOrdinalScale(padding = 0): ScaleOrdinal {
  const d3ScaleBand = scaleBand()
    .paddingInner(padding)
    .paddingOuter(0)
    .round(false);
  const scale = (value: string) => {
    return Number(d3ScaleBand(value));
  };
  const ordinalScale = scale as ScaleOrdinal;
  ordinalScale.domain = (domain?: string[]): string[] | ScaleOrdinal => {
    if (domain) {
      d3ScaleBand.domain(domain);
      return ordinalScale;
    }
    return d3ScaleBand.domain();
  };
  ordinalScale.range = (range?: [number, number]) => {
    if (range) {
      d3ScaleBand.range(range);
      return ordinalScale;
    }
    return d3ScaleBand.range();
  };
  ordinalScale.ticks = (): string[] => {
    return d3ScaleBand.domain();
  };
  ordinalScale.bandwidth = (): number => {
    return d3ScaleBand.bandwidth();
  };
  return ordinalScale;
}

/**
 * Return a continuous scale
 * @param type The type of scale (linear, log, sqrt)
 * @param domain The domain of the ordinal scale
 * @param minRange The max range of the scale
 * @param maxRange The min range of the scale
 * @param clamp if true, create a clamped scale
 */
export function createContinuousScale(
  type: ScaleContinuousTypes,
  domain: number[],
  minRange: number,
  maxRange: number,
  clamp = false,
): ScaleContinuousNumeric<number, number> {
  const scale = SCALES[type]();
  scale.domain(domain);
  scale.range([minRange, maxRange]);
  scale.clamp(clamp);
  return scale;
}

/**
 * Return a ScaleOrdinal
 * @param domain The domain of the ordinal scale
 * @param minRange The max range of the scale
 * @param maxRange The min range of the scale
 */
export function createOrdinalScale(
  domain: string[],
  minRange: number,
  maxRange: number,
  padding?: number,
): ScaleOrdinal {
  const scale = buildOrdinalScale(padding);
  scale.domain(domain);
  scale.range([minRange, maxRange]);
  return scale;
}

/**
 * Instanciate an ordinal scale function
 * @param scale The ordinal Scale
 * @param accessor The datum accessor function
 * @returns {ScaleFunction} The ScaleFunction
 */
export function getOrdinalScaleFn(
  scale: ScaleOrdinal,
  accessor: AccessorFn,
  centering = false,
): ScaleFunction {
  return (point: any) => {
    return scale(accessor(point)) + (centering ? scale.bandwidth() / 2 : 0);
  };
}

/**
 * Instanciate an continuous scale function
 * @param type {ScaleContinuousTypes} The type of continuous scale
 * @param domain The domain of the scale
 * @param accessor The accessor of the datum
 * @param minRange The min range of the scale
 * @param maxRange The max range of the scale
 * @param clamp if true, create a clamped scale
 */
export function getContinuousScaleFn(
  type: ScaleContinuousTypes,
  domain: number[],
  accessor: (point: any) => any,
  minRange: number,
  maxRange: number,
  clamp: boolean = false,
): ScaleFunction {
  const scale = createContinuousScale(type, domain, minRange, maxRange, clamp);
  return (point: any) => {
    return scale(accessor(point));
  };
}
