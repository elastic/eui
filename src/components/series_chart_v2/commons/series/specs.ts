import { Accessor } from '../data_ops/accessor';
import { ScaleType } from '../data_ops/scales';
import { AxisId, GroupId, SpecId } from '../ids';

export type Datum = any;
export type Rotation = 0 | 90 | -90 | 180;
export enum DataSeriesType {
  Bar = 'bar',
  Line = 'line',
  Area = 'area',
  Point = 'point',
}

export interface DataSeriesSpec {
  id: SpecId;
  groupId: GroupId;
  type: DataSeriesType;
  data: Datum[];
  scaleToExtent: boolean;
  xAccessor: Accessor;
  yAccessor: Accessor | Accessor[];
  xScaleType: ScaleType;
  yScaleType: ScaleType;
  stackAccessor?: Accessor | Accessor[];
}

/**
 * A Bar series spec. This specs specify how the bar series is configured
 * and provides data together with all configured accessors.
 */
export interface BasicSeriesSpec {
  id: SpecId;
  groupId: GroupId;
  data: Datum[];
  splitChartAccessor?: Accessor | Accessor[];
}

export interface BarSeriesSpec extends BasicSeriesSpec {
  xAccessor: Accessor;
  yAccessors: Accessor[];
  xScaleType: ScaleType;
  yScaleType: ScaleType;
  /** if true, the min y value is set to the minimum domain value, 0 otherwise */
  yScaleToDataExtent: boolean;
  splitSeriesAccessors?: Accessor[];
  stackAccessors?: Accessor[];
  colorAccessors?: Accessor[];
  tooltipLevel?: number;
}

export interface LineSeriesSpec extends BasicSeriesSpec {
  xAccessor: Accessor;
  yAccessors: Accessor[];
  xScaleType: ScaleType;
  yScaleType: ScaleType;
  /** if true, the min y value is set to the minimum domain value, 0 otherwise */
  yScaleToDataExtent: boolean;
  splitSeriesAccessors?: Accessor[];
  stackAccessors?: Accessor[];
  colorAccessors?: Accessor[];
  tooltipLevel?: number;
}

export interface AreaSeriesSpec extends BasicSeriesSpec {
  xAccessor: Accessor;
  yAccessors: Accessor[];
  xScaleType: ScaleType;
  yScaleType: ScaleType;
  /** if true, the min y value is set to the minimum domain value, 0 otherwise */
  yScaleToDataExtent: boolean;
  splitSeriesAccessors?: Accessor[];
  stackAccessors?: Accessor[];
  colorAccessors?: Accessor[];
  tooltipLevel?: number;
}

export interface AxisSpec {
  id: AxisId;
  groupId: GroupId;
  hide: boolean;
  showOverlappingTicks: boolean;
  showOverlappingLabels: boolean;
  position: AxisPosition;
  orientation: AxisOrientation;
  tickSize: number;
  tickPadding: number;
  tickFormat: (value: any) => string;
  /** Select the level of ticks to show. Currently not implemented. */
  groupingLevel?: number;
  /** The axis title */
  title?: string;
}

export enum AxisPosition {
  Top = 'top',
  Bottom = 'bottom',
  Left = 'left',
  Right = 'right',
}

export enum AxisOrientation {
  Vertical = 'vertical',
  Horizontal = 'horizontal',
}
