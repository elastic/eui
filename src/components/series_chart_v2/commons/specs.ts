import { Accessor } from './domain';
import { AxisId, GroupId, SpecId } from './ids';
import { ScaleType } from './scales';

type Datum = any;
export enum DataSeriesType {
  Bar = 'bar',
  Line = 'line',
  Area = 'area',
  Point = 'point',
}
export interface DataSeriesSpec {
  id: SpecId;
  groupId: GroupId;
  type: DataSeriesType;
  data: Datum[];
  scaleToExtent: boolean;
  xAccessor: Accessor;
  yAccessor: Accessor;
  xScaleType: ScaleType;
  yScaleType: ScaleType;
  /** An array of accessor for y value grouping/clustering */
  groupAccessors: Accessor[];
}

export interface AxisSpec {
  id: AxisId;
  groupId: GroupId;
  hide: boolean;
  showOverlappingTicks: boolean;
  showOverlappingLabels: boolean;
  position: AxisPosition;
  orientation: AxisOrientation;
  tickSize: number;
  tickPadding: number;
  tickFormat: (value: any) => string;
}

export enum AxisPosition {
  Top = 'top',
  Bottom = 'bottom',
  Left = 'left',
  Right = 'right',
}

export enum AxisOrientation {
  Vertical = 'vertical',
  Horizontal = 'horizontal',
}
