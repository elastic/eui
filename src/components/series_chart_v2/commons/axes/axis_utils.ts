import { max } from 'd3-array';
import { Domain, SpecDomain, SpecDomains } from '../data_ops/domain';
import { createContinuousScale, createOrdinalScale, ScaleType } from '../data_ops/scales';
import { Dimensions, Margins } from '../dimensions';
import { AxisId } from '../ids';
import { AxisOrientation, AxisPosition, AxisSpec, Rotation } from '../series/specs';
import { ChartConfig, ScalesConfig } from '../themes/theme';
import { BBoxCalculator } from './bbox_calculator';

export interface AxisTick {
  value: number | string;
  label: string;
  position: number;
}

export interface AxisTicksDimensions {
  axisScaleType: ScaleType;
  axisScaleDomain: Domain;
  tickValues: string[] | number[];
  ticksDimensions: Array<{ width: number; height: number }>;
  tickLabels: string[];
  maxTickWidth: number;
  maxTickHeight: number;
}

/**
 * Compute the ticks values and identify max width and height of the labels
 * so we can compute the max space occupied by the axis component.
 * @param axisSpec tbe spec of the axis
 * @param specDomains the domain associated
 * @param bboxCalculator an instance of the boundingbox calculator
 */
export function computeAxisTicksDimensions(
  axisSpec: AxisSpec,
  specDomains: SpecDomains,
  bboxCalculator: BBoxCalculator,
  scalesConfig: ScalesConfig,
  chartRotation: Rotation,
): AxisTicksDimensions {
  let tickValues: string[] | number[];
  let tickLabels: string[];
  let axisScaleType: ScaleType;
  let axisScaleDomain: Domain;
  const level = axisSpec.groupingLevel ? axisSpec.groupingLevel : 0;
  const verticalDomain = chartRotation === 0 ? specDomains.yDomain : specDomains.xDomains[level];
  const horizontalDomain = chartRotation === 0 ? specDomains.xDomains[level] : specDomains.yDomain;
  const axisDomain = axisSpec.orientation === AxisOrientation.Vertical ? verticalDomain : horizontalDomain;
  const verticalTicks = computeTicks(axisDomain, axisSpec, scalesConfig);
  tickValues = verticalTicks.tickValues;
  tickLabels = verticalTicks.tickLabels;
  axisScaleType = verticalTicks.axisScaleType!;
  axisScaleDomain = verticalTicks.axisScaleDomain!;
  // compute the boundingbox for each formatted label
  const ticksDimensions = tickLabels
    .map((tickLabel: string) => {
      const bbox = bboxCalculator.compute(tickLabel).getOrElse({
        width: 0,
        height: 0,
      });
      return {
        width: Math.ceil(bbox.width),
        height: Math.ceil(bbox.height),
      };
    })
    .filter((d) => d);
  const maxTickWidth = max(ticksDimensions, (bbox) => bbox.width) || 0;
  const maxTickHeight = max(ticksDimensions, (bbox) => bbox.height) || 0;
  return {
    axisScaleType,
    axisScaleDomain,
    ticksDimensions,
    tickValues,
    tickLabels,
    maxTickWidth,
    maxTickHeight,
  };
}

function computeTicks(
  specDomain: SpecDomain,
  axisSpec: AxisSpec,
  scalesConfig: ScalesConfig,
): {
  axisScaleType: ScaleType,
  axisScaleDomain: Domain,
  tickValues: any[],
  tickLabels: string[],
} {
  const { domain, scaleType } = specDomain;

  let tickValues: string[] | number[];
  let tickLabels: string[];

  if (scaleType === ScaleType.Ordinal) {
    const scale = createOrdinalScale(domain as string[], 1, 0, scalesConfig.ordinal.padding);
    tickValues = scale.ticks();
    tickLabels = tickValues.map(axisSpec.tickFormat);
  } else {
    const scale = createContinuousScale(scaleType, domain as number[], 1, 0);
    tickValues = scale.ticks();
    tickLabels = tickValues.map(axisSpec.tickFormat);
  }
  return {
    axisScaleType: scaleType,
    axisScaleDomain: domain,
    tickValues,
    tickLabels,
  };
}

export function getAvailableTicks(
  chartDimensions: Dimensions,
  axisSpec: AxisSpec,
  axisDimension: AxisTicksDimensions,
  scalesConfig: ScalesConfig,
) {
  const { width, height } = chartDimensions;
  const { axisScaleType, axisScaleDomain } = axisDimension;
  let allTicks: AxisTick[] = [];
  const minRange = axisSpec.orientation === AxisOrientation.Vertical ? height : 0;
  const maxRange = axisSpec.orientation === AxisOrientation.Vertical ? 0 : width;
  if (axisScaleType === ScaleType.Ordinal) {
    const scale = createOrdinalScale(
      axisScaleDomain as string[],
      minRange,
      maxRange,
      scalesConfig.ordinal.padding,
    );
    const ticks = scale.ticks();
    const bandwidth = scale.bandwidth();
    allTicks = ticks.map((tick) => {
      return {
        value: tick,
        label: axisSpec.tickFormat(tick),
        position: scale(tick) + bandwidth / 2,
      };
    });
  } else {
    const scale = createContinuousScale(
      axisScaleType,
      axisScaleDomain as number[],
      minRange,
      maxRange,
    );
    const ticks = scale.ticks();
    allTicks = ticks.map((tick) => {
      return {
        value: tick,
        label: axisSpec.tickFormat(tick),
        position: scale(tick),
      };
    });
  }
  return allTicks;
}
export function getVisibleTicks(
  allTicks: AxisTick[],
  axisSpec: AxisSpec,
  axisDim: AxisTicksDimensions,
  chartDimensions: Dimensions,
): AxisTick[] {
  const { orientation, showOverlappingTicks, showOverlappingLabels } = axisSpec;
  const { maxTickHeight, maxTickWidth } = axisDim;
  const { height } = chartDimensions;
  let firstTickPosition;
  if (axisSpec.orientation === AxisOrientation.Vertical) {
    firstTickPosition = allTicks[allTicks.length - 1].position;
  } else {
    firstTickPosition = allTicks[0].position;
  }

  let previousOccupiedSpace = firstTickPosition;
  const visibleTicks = [];
  for (let i = 0; i < allTicks.length; i++) {
    const { position } = allTicks[i];
    const requiredSpace =
      orientation === AxisOrientation.Vertical ? maxTickHeight / 2 : maxTickWidth / 2;
    const relativeTickPosition =
      orientation === AxisOrientation.Vertical ? height - position : position;
    if (i === 0) {
      visibleTicks.push(allTicks[i]);
      previousOccupiedSpace = firstTickPosition + requiredSpace;
    } else if (relativeTickPosition - requiredSpace >= previousOccupiedSpace) {
      visibleTicks.push(allTicks[i]);
      previousOccupiedSpace = relativeTickPosition + requiredSpace;
    } else {
      // still add the tick but without a label
      if (showOverlappingTicks || showOverlappingLabels) {
        const overlappingTick = {
          ...allTicks[i],
          label: showOverlappingLabels ? allTicks[i].label : '',
        };
        visibleTicks.push(overlappingTick);
      }
    }
  }
  return visibleTicks;
}

export function getAxisPosition(
  chartDimensions: Dimensions,
  chartMargins: Margins,
  axisSpec: AxisSpec,
  axisDim: AxisTicksDimensions,
  cumTopSum: number,
  cumBottomSum: number,
  cumLeftSum: number,
  cumRightSum: number,
) {
  const { top, left, height, width } = chartDimensions;
  const dimensions = {
    top,
    left,
    width,
    height,
  };
  let topIncrement = 0;
  let bottomIncrement = 0;
  let leftIncrement = 0;
  let rightIncrement = 0;

  if (axisSpec.orientation === AxisOrientation.Vertical) {
    const specLeft =
      axisSpec.position === AxisPosition.Left
        ? axisDim.maxTickWidth + cumLeftSum + chartMargins.left
        : left + width + cumRightSum;
    if (axisSpec.position === AxisPosition.Left) {
      leftIncrement =
        axisDim.maxTickWidth + axisSpec.tickSize + axisSpec.tickPadding + chartMargins.left;
    } else {
      rightIncrement =
        axisDim.maxTickWidth + axisSpec.tickSize + axisSpec.tickPadding + chartMargins.right;
    }
    dimensions.left = specLeft;
    dimensions.width = axisDim.maxTickWidth;
  } else {
    const specTop =
      axisSpec.position === AxisPosition.Top
        ? cumTopSum + chartMargins.top
        : top + height + cumBottomSum;
    if (axisSpec.position === AxisPosition.Top) {
      topIncrement =
        axisDim.maxTickHeight + axisSpec.tickSize + axisSpec.tickPadding + chartMargins.top;
    } else {
      bottomIncrement =
        axisDim.maxTickHeight + axisSpec.tickSize + axisSpec.tickPadding + chartMargins.bottom;
    }
    dimensions.top = specTop;
    dimensions.height = axisDim.maxTickHeight;
  }
  return { dimensions, topIncrement, bottomIncrement, leftIncrement, rightIncrement };
}

export function getAxisTicksPositions(
  chartDimensions: Dimensions,
  chartConfig: ChartConfig,
  scalesConfig: ScalesConfig,
  axisSpecs: Map<AxisId, AxisSpec>,
  axisDimensions: Map<AxisId, AxisTicksDimensions>,
) {
  const axisPositions: Map<AxisId, Dimensions> = new Map();
  const axisVisibleTicks: Map<AxisId, AxisTick[]> = new Map();
  const axisTicks: Map<AxisId, AxisTick[]> = new Map();
  let cumTopSum = 0;
  let cumBottomSum = 0;
  let cumLeftSum = 0;
  let cumRightSum = 0;
  axisDimensions.forEach((axisDim, id) => {
    const axisSpec = axisSpecs.get(id);
    if (!axisSpec) {
      return;
    }
    const allTicks = getAvailableTicks(chartDimensions, axisSpec, axisDim, scalesConfig);
    const visibleTicks = getVisibleTicks(allTicks, axisSpec, axisDim, chartDimensions);
    const axisPosition = getAxisPosition(
      chartDimensions,
      chartConfig.margins,
      axisSpec,
      axisDim,
      cumTopSum,
      cumBottomSum,
      cumLeftSum,
      cumRightSum,
    );
    cumTopSum += axisPosition.topIncrement;
    cumBottomSum += axisPosition.bottomIncrement;
    cumLeftSum += axisPosition.leftIncrement;
    cumRightSum += axisPosition.rightIncrement;
    axisPositions.set(id, axisPosition.dimensions);
    axisVisibleTicks.set(id, visibleTicks);
    axisTicks.set(id, allTicks);
  });
  return {
    axisPositions,
    axisTicks,
    axisVisibleTicks,
  };
}
