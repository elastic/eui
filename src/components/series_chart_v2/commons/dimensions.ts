import { AxisDimensions } from '../state/axis_utils';
import { AxisId } from './ids';
import { AxisOrientation, AxisPosition, AxisSpec } from './specs';

export interface Dimensions {
  top: number;
  left: number;
  width: number;
  height: number;
}
/**
 * Compute the chart dimension padding the parent dimension by the specified set of axis
 * @param parentDimensions the parent dimension
 * @param axisDimensions the axis dimensions
 * @param axisSpecs the axis specs
 */
export function computeChartDimensions(
  parentDimensions: Dimensions,
  axisDimensions: Map<AxisId, AxisDimensions>,
  axisSpecs: Map<AxisId, AxisSpec>,
): Dimensions {
  let vLeftAxisSpecWidth = 0;
  let vRightAxisSpecWidth = 0;
  let hTopAxisSpecHeight = 0;
  let hBottomAxisSpecHeight = 0;

  axisDimensions.forEach(({ maxTickWidth = 0, maxTickHeight = 0 }, id) => {
    const axisSpec = axisSpecs.get(id);
    if (!axisSpec) {
      return;
    }
    const { orientation, position, tickSize, tickPadding } = axisSpec;
    if (orientation === AxisOrientation.Horizontal) {
      if (position === AxisPosition.Top) {
        hTopAxisSpecHeight += maxTickHeight + tickSize + tickPadding;
      } else if (position === AxisPosition.Bottom) {
        hBottomAxisSpecHeight += maxTickHeight + tickSize + tickPadding;
      }
    } else {
      if (position === AxisPosition.Left) {
        vLeftAxisSpecWidth += maxTickWidth + tickSize + tickPadding;
      } else if (position === AxisPosition.Right) {
        vRightAxisSpecWidth += maxTickWidth + tickSize + tickPadding;
      }
    }
  });

  const chartWidth = parentDimensions.width - vLeftAxisSpecWidth - vRightAxisSpecWidth;
  const chartHeight = parentDimensions.height - hTopAxisSpecHeight - hBottomAxisSpecHeight;

  return {
    top: hTopAxisSpecHeight,
    left: vLeftAxisSpecWidth,
    width: chartWidth,
    height: chartHeight,
  };
}
