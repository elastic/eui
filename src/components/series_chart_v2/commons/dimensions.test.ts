import { AxisTicksDimensions } from './axes/axis_utils';
import { ScaleType } from './data_ops/scales';
import { computeChartDimensions, Margins } from './dimensions';
import { AxisId, getAxisId, getGroupId } from './ids';
import { AxisOrientation, AxisPosition, AxisSpec } from './series/specs';

describe('Computed chart dimensions', () => {
  const parentDim = {
    width: 100,
    height: 100,
    top: 0,
    left: 0,
  };
  const chartMargins: Margins = {
    left: 10,
    right: 10,
    top: 10,
    bottom: 10,
  };

  const axis1Dims = {
    axisScaleType: ScaleType.Linear,
    axisScaleDomain: [0, 1],
    tickValues: [0, 1],
    ticksDimensions: [{ width: 50, height: 20 }, { width: 50, height: 20 }],
    tickLabels: ['first', 'second'],
    maxTickWidth: 50,
    maxTickHeight: 20,
  };
  const axis1Spec = {
    id: getAxisId('axis_1'),
    groupId: getGroupId('group_1'),
    hide: false,
    showOverlappingTicks: false,
    showOverlappingLabels: false,
    position: AxisPosition.Left,
    orientation: AxisOrientation.Vertical,
    tickSize: 10,
    tickPadding: 10,
    tickFormat: (value: any) => {
      return `${value}`;
    },
  };
  test('should be equal to parent dimension with no axis minus margins', () => {
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    const chartDimensions = computeChartDimensions(parentDim, chartMargins, axisDims, axisSpecs);
    const expectedChartDimensions = { top: 10, left: 10, width: 80, height: 80 };
    expect(chartDimensions).toEqual(expectedChartDimensions);
  });
  test('should be padded by a left axis', () => {
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    axisDims.set(getAxisId('axis_1'), axis1Dims);
    axisSpecs.set(getAxisId('axis_1'), axis1Spec);
    const chartDimensions = computeChartDimensions(parentDim, chartMargins, axisDims, axisSpecs);
    const expectedChartDimensions = { top: 10, left: 80, width: 10, height: 80 };
    expect(chartDimensions).toEqual(expectedChartDimensions);
  });
  test('should be padded by a right axis', () => {
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    axisDims.set(getAxisId('axis_1'), axis1Dims);
    axisSpecs.set(getAxisId('axis_1'), { ...axis1Spec, position: AxisPosition.Right });
    const chartDimensions = computeChartDimensions(parentDim, chartMargins, axisDims, axisSpecs);
    const expectedChartDimensions = { top: 10, left: 10, width: 10, height: 80 };
    expect(chartDimensions).toEqual(expectedChartDimensions);
  });
  test('should be padded by a top axis', () => {
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    axisDims.set(getAxisId('axis_1'), axis1Dims);
    axisSpecs.set(getAxisId('axis_1'), {
      ...axis1Spec,
      position: AxisPosition.Top,
      orientation: AxisOrientation.Horizontal,
    });
    const chartDimensions = computeChartDimensions(parentDim, chartMargins, axisDims, axisSpecs);
    const expectedChartDimensions = { top: 50, left: 10, width: 80, height: 40 };
    expect(chartDimensions).toEqual(expectedChartDimensions);
  });
  test('should be padded by a bottom axis', () => {
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    axisDims.set(getAxisId('axis_1'), axis1Dims);
    axisSpecs.set(getAxisId('axis_1'), {
      ...axis1Spec,
      position: AxisPosition.Bottom,
      orientation: AxisOrientation.Horizontal,
    });
    const chartDimensions = computeChartDimensions(parentDim, chartMargins, axisDims, axisSpecs);
    const expectedChartDimensions = { top: 10, left: 10, width: 80, height: 40 };
    expect(chartDimensions).toEqual(expectedChartDimensions);
  });
});
