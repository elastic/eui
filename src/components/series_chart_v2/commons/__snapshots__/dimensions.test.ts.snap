// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Computed chart dimensions should be equal to parent dimension with no axis minus margins 1`] = `
Object {
  "height": 60,
  "left": 20,
  "top": 20,
  "width": 60,
}
`;

exports[`Computed chart dimensions should be padded by a bottom axis 1`] = `
Object {
  "height": 30,
  "left": 20,
  "top": 20,
  "width": 60,
}
`;

exports[`Computed chart dimensions should be padded by a left axis 1`] = `
Object {
  "height": 60,
  "left": 50,
  "top": 20,
  "width": 30,
}
`;

exports[`Computed chart dimensions should be padded by a right axis 1`] = `
Object {
  "height": 60,
  "left": 20,
  "top": 20,
  "width": 30,
}
`;

exports[`Computed chart dimensions should be padded by a top axis 1`] = `
Object {
  "height": 30,
  "left": 20,
  "top": 50,
  "width": 60,
}
`;
